lib/resty/etcd/v3.lua

```diff
@@ -28,7 +29,54 @@ local mt = { __index = _M }
 -- define local refresh function variable
 local refresh_jwt_token
 
-local function _request_uri(self, method, uri, opts, timeout, ignore_auth)
+
+local function incr(key, shm_name, failure_window)
+    local new_value, err, forcible = ngx_shared[shm_name]:incr(key, 1, 0, failure_window)
+    if err then
+        return nil, err
+    end
+
+    if forcible then
+        utils.log_warn("shared dict: ", shm_name, " is full, valid items forcibly overwritten")
+    end
+    return new_value, nil
+end
+
+
+local function restore(disable_duration, endpoint)
+    local ok, err = ngx_timer_at(disable_duration, function(premature)
+        if premature then
+            return
+        end
+
+        utils.log_info("restore an endpoint to health: ", endpoint.http_host)
+        endpoint.health_status = 1
+    end)
+
+    if not ok then
+        utils.log_error("failed to start timer to restore etcd endpoint to health: ", err)
+    end
+end
+
+
+local function report_failure(self, endpoint)
+    utils.log_info("report an endpoint failure: ", endpoint.http_host)
+    local failure_count, err = incr(endpoint.http_host, self.shm_name, self.failure_window)
+    if err then
+        utils.log_error("failed to incr etcd endpoint fail times: ", err)
+        return
+    end
+
+    -- only trigger once
+    if failure_count == self.failure_times then
```

Comment:
I think `>=` is better here.

Reply:
this is the point I need to discuss. if use `>=`, the `restore` may trigger many times,  I have no better ways to control trigger  `restore` once.

Reply:
let me think ablou it, fix later.

Reply:
I am inclined to the first way, i.e. workers are independent about the health state of etcd endpoints. are just restoring them independently.

Reply:
> It uses shm to share the failure states but only restoring the endpoint's health state in current worker.
> 
I think the health state in shm is belong to one endpoint, and shared by all workers.

when a worker run into `failure_count == self.failure_times`, it means this endpoint has trigger the `max_fails` in this worker, and in this worker, this endpoint is marked as unhealthy and prepare to restore. 

indeed, other workers are not aware of what happen occurring in this worker, but other workers do not need to be aware that, when once more `repore_fault`  occurs in other workers, they will do the same things. in other words, if there is no once more `repore_fault`  occurs in other workers, other workers will consider the endpoint is healthy and nothing to do.

in this mechanism,  all of workers will exhibit the same behavior. 

is right?

> We should either storing failure states inside LuaJIT and restoring it per worker or sharing failure states among workers, in the latter case, we should also save the state in the shm.

storing failure states inside LuaJIT means use `local` var? 

I think the first way will use more fault to confirm the same endpoint as unhealthy. If config the `max_fails` = 3, and the workers unmber is 3, they will use 9(3 * 3) faults to confirm the same endpoint is unhealthy, and the  `max_fails` lost its true meaning.

Reply:
you are right, I made the mistake of  `failure_count == self.failure_times`, it will trigger only in one worker. let me think about it.

Reply:
![131607855057_ pic](https://user-images.githubusercontent.com/30819887/102009194-c6732b00-3d70-11eb-8ef3-2f35f1293231.jpg)

now, the failure states shared in per worker, store in lua_shared_dict with ngx.worker.id, and trigger by different etcd client configurations.